---
title: "scmap-guide"
author: "Sean Wilson"
date: "2/15/2019"
output: 
  html_document:
    toc: true
params:
  d: !r Sys.Date()
  
---

Today's date is `r params$d`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Introduction

Template for using scmap to assign cell types to single cell RNAseq transcriptional data. The goal of this document is to provide an editable template showing a workflow to download datasets from GEO and use these to compare to a personally generated dataset. The package scmap (citation) can compare your single cell transcriptomes to previously annotated datsets and assign cell type and cluster information based on this.

# Preparation

First we need to load libraries required for the analysis. 

```{r libraries, include=T}
library(scmap) # the scmap package
library(SingleCellExperiment) # single cell experiment is the file type scmap uses for transcriptional data
library(GEOquery) # package to download GEO accesion data
library(Seurat) # package that allows for the analysis of single cell data, use to do QC and clustering and convert to sce post analysis
library(clustree) # clustering visualisation 
library(tidyverse) # essential data management package
```

There are two files we need to initially load into our run: the annotated dataset and test dataset. For this run, the annotated dataset is the single cell RNAseq Human Fetal Kidney dataset from Lindstrom (citation) located at GSE102596. The second is a combined dataset of two organoids generated from the same line, different batches and different ages, previously compared in our lineage tracing paper (citation).

```{r datasets}
getGEOSuppFiles(GEO = "GSE102596") # downloads the file from GEO into the working directory
untar("GSE102596/GSE102596_RAW.tar", exdir = "GSE102596/") # untar (unpacks) file
hfk.data <- read.table("GSE102596/GSM2741551_count-table-human16w.tsv.gz", sep = "\t") # converts the expression matrix into a data.table

reporters <- readRDS("/group/kidn1/Group-Little_MCRI/GROUPS/Stem cells and Regeneration/Single Cell/Reporters_18vs25Days/output/seurat/D18vsD25_combined_20clusters_seurat.rds") # loads previously prepared .rds file of organoid data
```

## Annotation

The annotated datset is downloaded, however it isn't actually annotated at this stage. As such, we need to cluster and annotate the dataset, and we will use Seurat to do this.

```{r hfk-seurat}
hfk <- CreateSeuratObject(raw.data = hfk.data, project = "Human_Kidney_hfk")

hfk <- NormalizeData(hfk)

hfk <- FindVariableGenes(object = hfk, mean.function = ExpMean, dispersion.function = LogVMR, 
                           x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)
hfk <- ScaleData(hfk, display.progress = F)
hfk <- RunPCA(object = hfk, pc.genes = hfk@var.genes, do.print = F)
#PCAPlot(object = geo, dim.1 = 1, dim.2 = 2)
PCElbowPlot(object = hfk)
hfk <- FindClusters(object = hfk, reduction.type = "pca", dims.use = 1:10, 
                      resolution = seq(0, 2, 0.2), print.output = 0, save.SNN = TRUE)
hfk <- RunTSNE(hfk, reduction.use = "pca", dims.use = 1:10, 
                 do.fast = T)
clustree(hfk)
TSNEPlot(object = hfk, group.by = "res.0.6")
hfk <- SetIdent(hfk, ident.use = hfk@meta.data$res.0.6)

FeaturePlot(hfk, features.plot = c("TMEM213", "GATA3", "CD14", "TCF21"), cols.use = c("grey", "red"))

markers <- FindAllMarkers(hfk, min.pct = 0.1, logfc.threshold = 0.5)

markers.list <- lapply(0:(length(unique(hfk@ident))-1), function(x) {
  markers %>%
    dplyr::filter(cluster == x, p_val_adj < 0.05, avg_logFC > 0) %>%
    dplyr::arrange(-avg_logFC) %>%
    select(Gene = gene, LogFC = avg_logFC, pVal = p_val_adj)
})
```

Once this is completed, and you are happy with the clustering and annotation of these clusters, the following chunk will rename each cluster and apply that identifier to each cell.

```{r hfg-annotation}

current.cluster.ids <- c(0,1,2,3,4,5,6,7,8,9,10)
new.cluster.ids <- c("Stroma 1", "Stroma 2", "Neph Progenitors", "Stroma DCN", "Endothelium", "Stroma 3", 
                     "Macrophage", "Coll Duct/Dist Tubule", "Prox Tubule/Gloms", "Cell Cycle", "Blood")
hfk@ident <- plyr::mapvalues(hfk@ident,from = current.cluster.ids, to=new.cluster.ids)
```









