---
title: "scmap-guide"
author: "Sean Wilson"
date: "2/15/2019"
output: 
  html_document:
    toc: true
params:
  d: !r Sys.Date()
  
---

Today's date is `r params$d`

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

Template for using scmap to assign cell types to single cell RNAseq transcriptional data. The goal of this document is to provide an editable template showing a workflow to download datasets from GEO and use these to compare to a personally generated dataset. The package scmap <https://doi.org/10.1038/nmeth.4644> or github <https://github.com/hemberg-lab/scmap> can compare your single cell transcriptomes to previously annotated datsets and assign cell type and cluster information based on this.

# Preparation

First we need to load libraries required for the analysis. 

```{r libraries, echo=FALSE}
library(scmap) # the scmap package
library(SingleCellExperiment) # single cell experiment is the file type scmap uses for transcriptional data
library(GEOquery) # package to download GEO accesion data
library(Seurat) # package that allows for the analysis of single cell data, use to do QC and clustering and convert to sce post analysis
library(clustree) # clustering visualisation 
library(tidyverse) # essential data management package
source("scmap-functions.R") # some minor edits to functions
```

There are two files we need to initially load into our run: the annotated dataset and test dataset. For this run, the annotated dataset is the single cell RNAseq Human Fetal Kidney dataset from Lindstrom (citation) located at GSE102596. The second is a combined dataset of two organoids generated from the same line, different batches and different ages, previously compared in our lineage tracing paper (citation).

```{r datasets}
getGEOSuppFiles(GEO = "GSE102596") # downloads the file from GEO into the working directory
untar("GSE102596/GSE102596_RAW.tar", exdir = "GSE102596/") # untar (unpacks) file
hfk.data <- read.table("GSE102596/GSM2741551_count-table-human16w.tsv.gz", sep = "\t") # converts the expression matrix into a data.table

reporters <- readRDS("/group/kidn1/Group-Little_MCRI/GROUPS/Stem cells and Regeneration/Single Cell/Reporters_18vs25Days/output/seurat/D18vsD25_combined_20clusters_seurat.rds") # loads previously prepared .rds file of organoid data
```

## Annotation

The annotated datset is downloaded, however it isn't actually annotated at this stage. As such, we need to cluster and annotate the dataset, and we will use Seurat to do this.

```{r hfk-seurat}
hfk <- CreateSeuratObject(raw.data = hfk.data, project = "Human_Kidney_hfk")

hfk <- NormalizeData(hfk)

hfk <- FindVariableGenes(object = hfk, mean.function = ExpMean, dispersion.function = LogVMR, 
                           x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)
hfk <- ScaleData(hfk, display.progress = F)
hfk <- RunPCA(object = hfk, pc.genes = hfk@var.genes, do.print = F)
#PCAPlot(object = geo, dim.1 = 1, dim.2 = 2)
PCElbowPlot(object = hfk)
hfk <- FindClusters(object = hfk, reduction.type = "pca", dims.use = 1:10, 
                      resolution = seq(0, 2, 0.2), print.output = 0, save.SNN = TRUE)
hfk <- RunTSNE(hfk, reduction.use = "pca", dims.use = 1:10, 
                 do.fast = T)
clustree(hfk)
TSNEPlot(object = hfk, group.by = "res.0.6")
hfk <- SetIdent(hfk, ident.use = hfk@meta.data$res.0.6)

FeaturePlot(hfk, features.plot = c("GATA3", "COL3A1", "NPHS1", "EPCAM", "SIX1", "PAX2"), cols.use = c("grey", "red"), nCol = 3)

markers <- FindAllMarkers(hfk, min.pct = 0.1, logfc.threshold = 0.5)

markers.list <- lapply(0:(length(unique(hfk@ident))-1), function(x) {
  markers %>%
    dplyr::filter(cluster == x, p_val_adj < 0.05, avg_logFC > 0) %>%
    dplyr::arrange(-avg_logFC) %>%
    select(Gene = gene, LogFC = avg_logFC, pVal = p_val_adj)
})
```

Once this is completed, and you are happy with the clustering and annotation of these clusters, the following chunk will rename each cluster and apply that identifier to each cell.

```{r hfk-annotation}

current.cluster.ids <- c(0,1,2,3,4,5,6,7,8,9,10)
new.cluster.ids <- c("Stroma 1", "Stroma 2", "Neph Progenitors", "Stroma DCN", "Endothelium", "Stroma 3", 
                     "Macrophage", "Coll Duct/Dist Tubule", "Prox Tubule/Gloms", "Cell Cycle", "Blood")
hfk@ident <- plyr::mapvalues(hfk@ident,from = current.cluster.ids, to=new.cluster.ids)

TSNEPlot(hfk, group.by = "ident")
```

The scmap package requires the data to be in sce format. This is easy to do with the Convert() function

```{r sce-convert}
hfk.sce <- Convert(from = hfk, to = "sce")
```

## Feature selection

Select the most informative genes from the input dataset (highlighted in red on the graph). I found that 50-100 features gives a "good" result, i.e. the majority of cells are assigned to their original clusters, there aren't many unassigned. Can increase or decrease this as required.

```{r features}
rowData(hfk.sce)$feature_symbol <- rownames(hfk.sce)
hfk.sce <- hfk.sce[!duplicated(rownames(hfk.sce)),]
hfk.sce <- selectFeatures.SingleCellExperiment(hfk.sce, suppress_plot = F, n_features = 100)
```

## scmap-cluster

The scmap-cluster index of a reference dataset is created by finding the median gene expression for each cluster.

```{r scmap-cluster}
hfk.sce <- indexCluster.SingleCellExperiment(hfk.sce, cluster_col = "ident")
head(metadata(hfk.sce)$scmap_cluster_index)
heatmap(as.matrix(metadata(hfk.sce)$scmap_cluster_index))
```

## Projection analysis

We now have an index of median gene expression relating to each of the clusters from the input data. Now we can use this index to assign identity to an input dataset. To test this, we can project the index dataset back onto itself, then we can use a different dataset and see where those cells are assigned. The threshold of simmilarity required to assign a cell to a cluster can be adjusted.

```{r projection-test}
scmapCluster_results <- scmapCluster.SingleCellExperiment(projection = hfk.sce,
                                     index_list = list(hfk = metadata(hfk.sce)$scmap_cluster_index),
                                     threshold = 0.5)

# visualise by sankey graphing method
plot(
  getSankey(
    colData(hfk.sce)$ident, 
    scmapCluster_results$scmap_cluster_labs[,'hfk'],
    plot_height = 800,
    plot_width = 1200
  )
)

# Plot the similarity value for each cell

qplot(x = scmapCluster_results$scmap_cluster_siml[,'hfk'],
      binwidth = 0.01)


# TSNE plot of reassigned cells by cluster
hfk@meta.data$scmap <- scmapCluster_results$scmap_cluster_labs[,'hfk']
TSNEPlot(hfk, do.return=T, no.legend = F, do.label = T)
TSNEPlot(hfk, do.label =F, group.by = "scmap")


```

There is a small amount of shuffling of identities doing this analysis, the cluster most affected looks to be the nephron progenitors. This could be because there is a stronger score with a reduced number of features in cells closer to a committed phate instead of their progenitor state. For now, let's say that there is an acceptable assignment of cell types using our feature based index. Now, we can compare and assign a seperate dataset.

```{r projection-organoids}
reporters <- SetIdent(reporters, ident.use = reporters@meta.data$res.1.6)
TSNEPlot(reporters, do.label = T)

rep.sce <- Convert(from = reporters, to = "sce")
rowData(rep.sce)$feature_symbol <- rownames(rep.sce)
#rep.sce <- selectFeatures.SingleCellExperiment(rep.sce, n_features = 500, suppress_plot = F)
#rep.sce <- indexCluster.SingleCellExperiment(rep.sce, cluster_col = "res.1.6")
#heatmap(as.matrix(metadata(rep.sce)$scmap_cluster_index))

scmapCluster_results_rep <- scmapCluster.SingleCellExperiment(
  projection = rep.sce, 
  index_list = list(
    reporters = metadata(hfk.sce)$scmap_cluster_index
  ), threshold = 0.2
)

head(scmapCluster_results_rep$scmap_cluster_labs)
head(scmapCluster_results_rep$scmap_cluster_siml)

plot(
  getSankey(
    colData(rep.sce)$"res.1.6", 
    scmapCluster_results_rep$scmap_cluster_labs[,'reporters'],
    plot_height = 800,
    plot_width = 1200
  )
)

qplot(x = scmapCluster_results_rep$scmap_cluster_siml[,'reporters'], binwidth = 0.01)
reporters@meta.data$scmap <- scmapCluster_results_rep$scmap_cluster_labs[,'reporters']

# Plot scmap assigned cluster identity to tsne plot of data

TSNEPlot(reporters, do.return=T, no.legend = F, do.label = T)
TSNEPlot(reporters, do.label =T, group.by = "scmap")

```

Interestingly, there are two peaks in the distribution of alignment scores, one around 0.5 and one around 0.8. What cell types contain the lower aligment scores compared to the higher ones? My guess is that the multiple "stromal" populations contain similar scores for their feature index, and will those cells will be less specific to any one cluster.

```{r alignment-dist}
reporters@meta.data$scores <- round(scmapCluster_results_rep$scmap_cluster_siml[, 'reporters'], digits = 1) # round so as to make graphing easier
TSNEPlot(reporters, group.by = "scores")

```

The cell cycle cluster on the left and the stromal population left centre give very strong similarity to the corresponding index in the human fetal kidney data. The weakest regions of similarity are where the podocyte cells (top middle) and subset of a stromal cluster (lower middle right) are located. The HFK dataset doesn't have many podocytes, and there will be differences in stromal pops, so these don't seem out of place. there is a medium-high similarity in the majority of nephron lineages. I think this has done a decent enough job at assigning cell types based on a limited dataset index (the HFK), and the differences in experiments and cell lines (organoids generated from iPSCs vs real human fetal kidney samples).

This template can serve as a launch point for any analysis using scmap. I'll continue with our other datasets, also try and use other available datasets as an index, as well as using annotated organoids against eachother. I can also try to "merge" the two datasets together prior using a cca/multicca, subset individual datasets and repeat the process. This might eliminate more batch variation and cause higher correlation scores.

```{r session-info}
sessionInfo()
```












